docs/whitepaper/src/appendix.tex

```diff
@@ -122,49 +132,67 @@ \subsection{Algorithms}\label{subsec:algorithms}
+    \end{algorithmic}\label{alg:algorithm-2.0}
+\end{algorithm}
+
+\begin{algorithm}[ht]
+    \begin{algorithmic}[1]
+        \caption{$\textsf{IsValidBlock}(B, C_{text{\loc}})$}
+        \State \lstinline|// All indexes except the epoch index are omitted.|
+        \State Parse $B$ as ${(h, sl, \textsf{state}, v^{\text{vrf}}, \pi^{\text{sl}}, \sigma_{\text{KES}})}$ \lstinline|// value h is a block body hash.|
+        \State \lstinline|// Prepare constants:|
+
+        \State $(e_n, V_{\text{ver}}, \eta_{n}, \eta_{n - 2}, T^{\text{cons}}, T^{\text{lead}, T^{\text{sync}}) = \textsf{PrepareForBlockValidation}(sl, v^{\text{vrf}}, C_{\text{loc}})$
+        \State \lstinline|// Check consensus membership:|
+        \State Parse ${\pi^{\text{e}}}$ from $state$
+        \State Set ${r^{\text{e}} \leftarrow \pi^{\text{e}}\textsf{.extract\_random\_number()}}$
+        \State Set ${y^{\text{cons}} \leftarrow {\mathcal{H}(r^{\text{e}} || \textsf{CONS})}}$
+        \State Set ${\textsf{valid\_member} \leftarrow (v^{\text{vrf}} \in V_{\text{ver}}) \wedge (y^{\text{cons}} < T^{\text{cons}})}$.
+
+        \State \lstinline|// Check VRF proof:|
         \State Set $\textsf{valid\_proof} \leftarrow \mathcal{F}_{\text{VRF}}\textsf{.verify(}
-        v^{\text{vrf}}, m_{\text{hash}},\pi^{\text{sl}} \textsf{)}$
+        v^{\text{vrf}}, \mathcal{H}(\eta_n || sl),\pi^{\text{sl}} \textsf{)}$
 
-        \State \lstinline|// Check the leadership|
-        \State Set ${\pi^{\text{sl}*} \leftarrow \pi^{\text{sl}}\textsf{.extract\_random\_number()}}$
-        \State Set ${y^{\text{lead}} \leftarrow {\mathcal{H}(\pi^{\text{sl}*} || \textsf{LEAD})}}$
-        \State Set ${\textsf{valid\_leader} \leftarrow y^{\text{lead} < T^{\text{lead}, k}}}$.
+        \State \lstinline|// Check the leadership:|
+        \State Set ${r^{\text{sl}} \leftarrow \pi^{\text{sl}}\textsf{.extract\_random\_number()}}$
+        \State Set ${y^{\text{lead}} \leftarrow {\mathcal{H}(r^{\text{sl}} || \textsf{LEAD})}}$
+        \State Set ${\textsf{valid\_leader} \leftarrow y^{\text{lead}} < T^{\text{lead}}}$.
 
-        \State \lstinline|// Check KES signature|
-        \State Get $v^{\text{kes}}$ from $\textsf{state}$ by $v^{\text{vrf}}$
-        \State Set ${\pi^{\text{sl}*} \leftarrow \pi^{\text{sl}}\textsf{.extract\_random\_number()}}$
-        \State Set $m_{\text{hash}} \leftarrow \mathcal{H}(B)$
+        \State \lstinline|// Check KES signature:|
+        \State Set ${y^{\text{rand}} \leftarrow {\mathcal{H}(r^{\text{sl}} || \textsf{RAND})}}$
+        \State Parse $v^{\text{kes}}$ from $V_{\text{ver}}$
         \State Set $\textsf{valid\_signature} \leftarrow \mathcal{F}_{\text{KES}}\textsf{.verify(}
-        B,\pi^{\text{sl}}, \sigma_{\text{KES}}^{\text{sl}}, v^{\text{kes}}) \textsf{)}$
+        \mathcal{H}(h || \textsf{state} || sl || y^{\text{rand}}), \pi^{\text{sl}}, \sigma_{\text{KES}}, v^{\text{kes}}) \textsf{)}$
```

Comment:
Again probably state root hash should be used here.

Reply:
Yes sure, will move this into transformations inside KES functionality